mitiq/tests/test_folding_cirq.py

```diff
@@ -466,6 +467,102 @@ def test_fold_gates_at_random_seed_one_qubit():
     )
     assert _equal(folded, correct)
 
+    # Medium stretch, fold two gates
+    folded = fold_gates_at_random(circuit, stretch=2.5, seed=1)
+    correct = Circuit(
+        [ops.X.on(qubit)] ,
+        [ops.Y.on(qubit)] * 3,
+        [ops.Z.on(qubit)] * 3,
+    )
+    assert _equal(folded, correct)
+
+    # Max stretch, fold three gates
+    folded = fold_gates_at_random(circuit, stretch=3, seed=1)
+    correct = Circuit(
+        [ops.X.on(qubit)] * 3,
+        [ops.Y.on(qubit)] * 3,
+        [ops.Z.on(qubit)] * 3,
+    )
+    assert _equal(folded, correct)
+
+
+def test_fold_random_min_stretch():
+    """Tests that folding at random with min stretch returns a copy of the input circuit."""
+    # Test circuit
+    # 0: ───H───@───@───
+    #           │   │
+    # 1: ───H───X───@───
+    #               │
+    # 2: ───H───T───X───
+    qreg = LineQubit.range(3)
+    circ = Circuit(
+        [ops.H.on_each(*qreg)],
+        [ops.CNOT.on(qreg[0], qreg[1])],
+        [ops.T.on(qreg[2])],
+        [ops.TOFFOLI.on(*qreg)]
+    )
+
+    folded = fold_gates_at_random(circ, stretch=1, seed=1)
+    assert _equal(folded, circ)
+    assert folded is not circ
+
+
+def test_fold_random_max_stretch():
+    """Tests that folding at random with max stretch folds all gates on a multi-qubit circuit."""
+    # Test circuit
+    # 0: ───H───@───@───
+    #           │   │
+    # 1: ───H───X───@───
+    #               │
+    # 2: ───H───T───X───
+    qreg = LineQubit.range(3)
+    circ = Circuit(
+        [ops.H.on_each(*qreg)],
+        [ops.CNOT.on(qreg[0], qreg[1])],
+        [ops.T.on(qreg[2])],
+        [ops.TOFFOLI.on(*qreg)]
+    )
+
+    folded = fold_gates_at_random(circ, stretch=3, seed=1)
+    correct = Circuit(
+        [ops.H.on_each(*qreg)] * 3,
+        [ops.CNOT.on(qreg[0], qreg[1])] * 3,
+        [ops.T.on(qreg[2]), ops.T.on(qreg[2])**-1, ops.T.on(qreg[2])],
+        [ops.TOFFOLI.on(*qreg)] * 3
+    )
+    assert _equal(folded, correct)
+
+
+def test_fold_random_bad_stretch():
+    """Tests that an error is raised when a bad stretch is provided."""
+    with pytest.raises(ValueError):
+        fold_gates_at_random(Circuit(), stretch=4)
+
+
+def test_fold_random_no_repeats():
+    """Tests folding at random to ensure that no gates are folded twice and folded gates
+    are not folded again.
+    """
+    # Test circuit
+    # 0: ───H───@───@───
+    #           │   │
+    # 1: ───H───X───@───
+    #               │
+    # 2: ───H───T───X───
+    qreg = LineQubit.range(3)
+    circ = Circuit(
+        [ops.H.on_each(*qreg)],
+        [ops.CNOT.on(qreg[0], qreg[1])],
+        [ops.T.on(qreg[2])],
+        [ops.TOFFOLI.on(*qreg)]
+    )
+
+    for stretch in np.linspace(1., 3., 10):
+        folded = fold_gates_at_random(circ, stretch=stretch, seed=1)
+        print(folded, "\n\n")
```

Comment:
Should replace this with an automated test before merge. An automated but less comprehensive is better than a manual one as the manual ones will quickly outpace our ability to check.

Perhaps you could use total gate count for a given gate type to ensure none are folded too many times and that folds aren't reused?

Reply:
Completely agree. I used a circuit in which every gate only occurs once to check no gates are folded more than once, checking total gate count :)